common/ipmi/oem_1s_handler.c

```diff
@@ -0,0 +1,864 @@
+	} else if (msg->data[0] == 3) { // Set GPIO direction status
+		if (msg->data_len == 3) {
+			if (msg->data[2]) {
+				gpio_conf(gpio_num, GPIO_OUTPUT);
+			} else {
+				gpio_conf(gpio_num, GPIO_INPUT);
+			}
+			msg->data[0] = gpio_num;
+			msg->data[1] = msg->data[2];
+			completion_code = CC_SUCCESS;
+		}
+	}
+
+	if (completion_code != CC_SUCCESS) {
+		msg->data_len = 0;
+	} else {
+		msg->data_len = 2; // Return GPIO number, status
+	}
+	msg->completion_code = completion_code;
+	return;
+}
+
+__weak void OEM_1S_I2C_DEV_SCAN(ipmi_msg *msg)
+{
+	if ((msg->data[0] == 0x9C) && (msg->data[1] == 0x9C) && (msg->data[2] == 0x00)) {
+		while (1)
+			; // hold firmware for debug only
+	}
+
+	if (msg->data_len != 1) { // only input scan bus
+		msg->completion_code = CC_INVALID_LENGTH;
+		return;
+	}
+
+	uint8_t bus = i2c_bus_to_index[msg->data[0]];
+
+	i2c_scan(bus, &msg->data[0], (uint8_t *)&msg->data_len);
+
+	msg->completion_code = CC_SUCCESS;
+	return;
+}
+
+__weak void OEM_1S_GET_BIC_STATUS(ipmi_msg *msg)
+{
+	if (msg->data_len != 0) {
+		msg->completion_code = CC_INVALID_LENGTH;
+		return;
+	}
+
+	msg->data[0] = FIRMWARE_REVISION_1;
+	msg->data[1] = FIRMWARE_REVISION_2;
+
+	msg->data_len = 2;
+	msg->completion_code = CC_SUCCESS;
+	return;
+}
+
+__weak void OEM_1S_RESET_BIC(ipmi_msg *msg)
+{
+	if (msg->data_len != 0) {
+		msg->completion_code = CC_INVALID_LENGTH;
+		return;
+	}
+
+	submit_bic_warm_reset();
+
+	msg->data_len = 0;
+	msg->completion_code = CC_SUCCESS;
+	return;
+}
+
+__weak void OEM_1S_12V_CYCLE_SLOT(ipmi_msg *msg)
+{
+	if (msg->data_len != 0) {
+		msg->completion_code = CC_INVALID_LENGTH;
+		return;
+	}
+
+	int ret = pal_submit_12v_cycle_slot();
+	switch (ret) {
+	case SUCCESS_12V_CYCLE_SLOT:
+		msg->completion_code = CC_SUCCESS;
+		break;
+	case NOT_SUPPORT_12V_CYCLE_SLOT:
+		msg->completion_code = CC_INVALID_CMD;
+		break;
+	case SLOT_OFF_FAILED:
+	case SLOT_ON_FAILED:
+		msg->completion_code = CC_UNSPECIFIED_ERROR;
+		break;
+	default:
+		msg->completion_code = CC_UNSPECIFIED_ERROR;
+		break;
+	}
+
+	return;
+}
+
+void send_gpio_interrupt(uint8_t gpio_num)
+{
+	ipmb_error status;
+	ipmi_msg msg;
+	uint8_t gpio_val;
+
+	printf("Send gpio interrupt %d to BMC\n", gpio_num);
+	gpio_val = gpio_get(gpio_num);
+
+	msg.data_len = 5;
+	msg.InF_source = SELF;
+	msg.InF_target = BMC_IPMB;
+	msg.netfn = NETFN_OEM_1S_REQ;
+	msg.cmd = CMD_OEM_1S_SEND_INTERRUPT_TO_BMC;
+
+	msg.data[0] = IANA_ID & 0xFF;
+	msg.data[1] = (IANA_ID >> 8) & 0xFF;
+	msg.data[2] = (IANA_ID >> 16) & 0xFF;
+	msg.data[3] = gpio_num;
+	msg.data[4] = gpio_val;
+
+	status = ipmb_read(&msg, IPMB_inf_index_map[msg.InF_target]);
+	if (status == IPMB_ERROR_FAILURE) {
+		printf("Fail to post msg to txqueue for gpio %d interrupt\n", gpio_num);
+	} else if (status == IPMB_ERROR_GET_MESSAGE_QUEUE) {
+		printf("No response from bmc for gpio %d interrupt\n", gpio_num);
+	}
+}
+
+void IPMI_OEM_1S_handler(ipmi_msg *msg)
+{
+	switch (msg->cmd) {
```

Comment:
msg == NULL check?

Reply:
We have fixed it and done the forced update.
Please review it. Thanks.